peer/persistentpeer_mgr.go

```diff
@@ -0,0 +1,621 @@
+		// manager.
+		case false:
+			peerLog.Info(
+				"Removing conn req:", connReq.Addr.String(),
+			)
+			m.cfg.ConnMgr.Remove(connReq.ID())
+		}
+	}
+
+	peer.connReqs = updatedConnReqs
+	cancelChan := peer.getRetryCanceller()
+
+	// Any addresses left in addrMap are new ones that we have not made
+	// connection requests for. So create new connection requests for those.
+	// If there is more than one address in the address map, stagger the
+	// creation of the connection requests for those.
+	go func() {
+		ticker := time.NewTicker(multiAddrConnectionStagger)
+		defer ticker.Stop()
+
+		for _, addr := range addrMap {
+			// Send the persistent connection request to the
+			// connection manager, saving the request itself so we
+			// can cancel/restart the process as needed.
+			connReq := &connmgr.ConnReq{
+				Addr:      addr,
+				Permanent: true,
+			}
+
+			m.Lock()
+			peer.connReqs = append(peer.connReqs, connReq)
+			m.Unlock()
+
+			peerLog.Debugf("Attempting persistent connection to "+
+				"channel peer %v", addr)
+
+			go m.cfg.ConnMgr.Connect(connReq)
+
+			select {
+			case <-m.quit:
+				return
+			case <-cancelChan:
+				return
+			case <-ticker.C:
+			}
+		}
+	}()
+}
+
+// UnassignedConnID is the default connection ID that a request can have before
+// it actually is submitted to the connmgr.
+// TODO(conner): move into connmgr package, or better, add connmgr method for
+// generating atomic IDs.
+const UnassignedConnID uint64 = 0
+
+// CancelConnReqs stops all persistent connection requests for a given pubKey.
+// Any retry canceller channel is canceled first. Afterwards, each connection
+// request is removed from the connmgr. The caller can optionally specify a
+// connection ID to ignore, which prevents us from canceling a successful
+// request. All persistent connReqs for the provided pubKey are discarded after
+// the operation.
+func (m *PersistentPeerManager) CancelConnReqs(pubKey *btcec.PublicKey,
+	skip *uint64) {
+
+	m.Lock()
+	defer m.Unlock()
+
+	m.cancelConnReqsUnsafe(pubKey, skip)
+}
+
+// cancelConnReqsUnsafe cancels all the connection requests for a peer with the
+// given pubKey. An optional skip ID can be specified if any connReqs should be
+// ignored. This function will also cancel the retry canceler of the peer.
+// NOTE: that this is the method is not thread safe and should only be called if
+// the PersistentPeerManager mutex lock is held.
+func (m *PersistentPeerManager) cancelConnReqsUnsafe(pubKey *btcec.PublicKey,
+	skip *uint64) {
+
+	peer, ok := m.conns[route.NewVertex(pubKey)]
+	if !ok {
+		return
+	}
+
+	// First, cancel any lingering persistent retry attempts, which will
+	// prevent retries for any with backoffs that are still maturing.
+	peer.cancelRetries()
+
+	// Next, check to see if we have any outstanding persistent connection
+	// requests to this peer. If so, then we'll remove all of these
+	// connection requests, and also delete the entry from the map.
+	if len(peer.connReqs) == 0 {
```

Comment:
I think at this point, there may still be a `ConnectPeer` goroutine blocked on obtaining the lock. After this function returns and unblocks, `ConnectPeer` may still add something to the `connReqs`?

Reply:
Yes this seems possible, if the goroutine spawned by `ConnectPeer` doesn't get canceled by the chan and is waiting on the mutex lock